docs/basic/divide-and-conquer.md

```diff
@@ -1,88 +1,163 @@
+1. Divide: Decompose original problem into two or more similar sub-problems.
+2. Conquer: After decomposing problems into some easily solvable bound, solve the sub-problem recursively until solved.
+3. Combine: Combine results from sub-problems to obtain final results of original problem.
 
-显然有时候递归处理是高效的，比如归并排序， **有时候是低效的** ，比如数孙悟空身上的毛，因为堆栈会消耗额外空间，而简单的递推不会消耗空间。比如这个例子，给一个链表头，计算它的长度：
+Generally, problems which can be solved with divide and conquer paradigm have these features:
 
-```java
-/* 典型的递推遍历框架 */
-public int size(Node head) {
-    int size = 0;
-    for (Node p = head; p != null; p = p.next) size++;
-    return size;
-}
-/* 我偏要递归，万物皆递归 */
-public int size(Node head) {
-    if (head == null) return 0;
-    return size(head.next) + 1;
+- The problem can be easily solved when its size is down to a certain level.
+- The problem can be decomposed into several sub-problems with smaller size, i.e., the problem has the property of optimal substructure. Results of sub-problems decomposed from original problem can be used and combined into results of original problem.
+- Sub-problems decomposed from original problem are independent to each other, i.e., Sub-problems do not contain common sub-problems between them.
+
+???+warning "Warning"
+    If sub-problems are not independent, then algorithms using divide-and-conquer paradigm need to solve common sub-problems repeatedly. Thus, unnecessary works would be done. Under such circumstances it is better to use [dynamic programming](../dp/basic.md) rather than divide-and-conquer paradigm.
+
+Let take merge sort as an example. Assume we have a function named `merge_sort` as an implementation of merge sort. The job of this function is **to sort an input array**. The problem is obviously can be decomposed. Sorting an array is equal to sorting two halves of this array, then combine these to one.
+
+```cpp
+// This block of code has been simplified.
+void merge_sort(type* an_array) {
+  if (is_easy_to_process) return;
+  merge_sort(left_half);
+  merge_sort(right_half);
+  merge(left_half, right_half);
 }
 ```
 
-### 写递归的技巧
+If we call the function with half of the array, then after processing this half of array will become sorted. We can notice that `merge_sort` is similar with template of post-order traversal of binary tree. Like the three parts of divide and conquer mentioned before, firstly decompose left and right, then process their combination. Combining is unstacking, which is equal to post-order traversal.
+
+The implementation of function `merge` is same to implementation that combines two ordered lint table into one.
+
+## Key points
+
+### Keys of Writing Recursion
 
-我的一点心得是： **明白一个函数的作用并相信它能完成这个任务，千万不要试图跳进细节。** 千万不要跳进这个函数里面企图探究更多细节，否则就会陷入无穷的细节无法自拔，人脑能压几个栈啊。
+**Understand a function's purpose and believe it can finish the job. DO NOT jump into the function for discovering more details,** or you will cannot help yourself from getting out from the infinite details. <!---人脑能压几个栈啊。-->
```

Comment:
```suggestion
**Understand a function's purpose and believe it can finish the job. DO NOT jump into the function for discovering more details,** or you will may be unable to get out of the infinite details, since human is not so good at maintaining a stack in mind.
```
🤔  这样可吗

Reply:
可 thx